File: cascadingFailuresTaskMigration/src/main/java/MPFTM/CalculatePonField.java
package MPFTM;

import input.Group;
import input.PotentialField;
import input.Robot;
import main.Function;
import org.jgrapht.alg.interfaces.ShortestPathAlgorithm;
import org.jgrapht.graph.DefaultUndirectedWeightedGraph;
import org.jgrapht.graph.DefaultWeightedEdge;

import java.util.HashMap;
import java.util.Set;

public class CalculatePonField {
    private DefaultUndirectedWeightedGraph<Integer, DefaultWeightedEdge> arcGraph;
    private HashMap<Integer, Group> idToGroups;
    private HashMap<Integer,Robot> idToRobots;
    private ShortestPathAlgorithm<Integer, DefaultWeightedEdge> shortestPath ;
    Double a;
    Double b;
    private HashMap<Integer,Double> idToI ;
    final Double y = 0.005;
    final Double yn = 0.3;
    final Double xn = 0.1;
    final Double x = 0.01;

    public CalculatePonField(HashMap<Integer, Group> idToGroups, HashMap<Integer, Robot> idToRobots, DefaultUndirectedWeightedGraph<Integer, DefaultWeightedEdge> arcGraph,
                             HashMap<Integer,Double> idToI ,ShortestPathAlgorithm<Integer, DefaultWeightedEdge> shortestPath, Double a, Double b) {
        this.idToGroups = idToGroups;
        this.idToRobots = idToRobots;
        this.arcGraph = arcGraph;
        this.shortestPath = shortestPath;
        this.a = a;
        this.b = b;
        this.idToI = idToI;
    }

    public HashMap<Integer, PotentialField> calculateIntraP() {
        //计算组件的势场
        HashMap<Integer,PotentialField> IntraPotential = new HashMap<>();
        Double Isum = 0.0;


        for (Integer id : idToRobots.keySet()) {
            Isum+= idToI.get(id);
        }
        Double Imean = Isum/idToRobots.size();
        for (Integer id : idToRobots.keySet()) {
            Robot robot = idToRobots.get(id);
            PotentialField p = new PotentialField();
            //如何设置p
            //设置引力势场，可以加一个对(I-Imean)
            Double I = idToI.get(id);
            p.setPegra(-a*gain(I-Imean));

            //设置斥力势场
            Double ro = 0.0;
            //节点的相邻边
            Set<DefaultWeightedEdge> defaultWeightedEdges = arcGraph.edgesOf(id);
            for (DefaultWeightedEdge defaultWeightedEdge : defaultWeightedEdges) {
                Robot targetRobot = idToRobots.get(arcGraph.getEdgeTarget(defaultWeightedEdge));
                if (targetRobot.getRobotId()==id) {
                    targetRobot = idToRobots.get(arcGraph.getEdgeSource(defaultWeightedEdge));
                }
                if (targetRobot.getGroupId()!=robot.getGroupId()) {
                    continue;
                }
                if (targetRobot.getFaultA()==1) {
                    //到故障节点的距离成反比
                    ro+=1/arcGraph.getEdgeWeight(defaultWeightedEdge);
                }
            }
            if (robot.getFaultA()==1) {
                p.setPerep(Double.MAX_VALUE/2);
            } else if(ro!=0){
                p.setPerep(b*(y*1/ro)*(1/ro));
            } else {
                p.setPerep(0.0);
            }
            IntraPotential.put(id,p);

            //更新过载故障情况
            Function function = new Function(idToRobots,idToGroups);
            Double faultO = 1-function.calculateOverLoadIS(idToRobots.get(id));
            robot.setFaultO(faultO);

        }



        return IntraPotential;
    }

    public HashMap<Integer, PotentialField> calculateInterP() {
        //计算网络层的势场
        HashMap<Integer, PotentialField> InterPotential = new HashMap<>();
        for (Integer groupId : idToGroups.keySet()) {
            Group group = idToGroups.get(groupId);
            PotentialField p = new PotentialField();
            //计算网络层的引力势场（向势场势场降低的方向传递）
            p.setPegra(a*xn*(group.getGroupLoad()));
            //计算网络层中的斥力场
            int fk = 0;
            Set<Integer> robotIdInGroup = group.getRobotIdInGroup();
            for (Integer id : robotIdInGroup) {
                //计算group中的故障节点
                Robot robot = idToRobots.get(id);
                if (robot.getFaultA()==1) {
                    fk+=1;
                }
            }
            int nk = robotIdInGroup.size();
            if (fk== nk) {
                p.setPerep(Double.MAX_VALUE/2);
            } else {
                p.setPerep(b*(yn*(double)fk/(nk-fk)));
            }
            InterPotential.put(groupId,p);
        }

        return InterPotential;
    }

    private Double gain(double x) {

       //return (Math.exp(Math.log(x+1))-Math.exp(-Math.log(x+1)))/(Math.exp(Math.log(x+1))+Math.exp(-Math.log(x+1)));
        return x;
    }
}

File: cascadingFailuresTaskMigration_python/MPFTM/calculate_pon_field.py
import networkx as nx
from typing import Dict
from input.group import Group
from input.robot import Robot
from input.potential_field import PotentialField
from main.function import Function


class CalculatePonField:
    def __init__(self, id_to_groups: Dict[int, Group], id_to_robots: Dict[int, Robot],
                 arc_graph: nx.Graph, id_to_i: Dict[int, float], shortest_path_dict: Dict,
                 a: float, b: float):
        self.id_to_groups = id_to_groups
        self.id_to_robots = id_to_robots
        self.arc_graph = arc_graph
        self.shortest_path_dict = shortest_path_dict
        self.a = a
        self.b = b
        self.id_to_i = id_to_i
        self.y = 0.005
        self.yn = 0.3
        self.xn = 0.1
        self.x = 0.01

    def calculate_intra_p(self) -> Dict[int, PotentialField]:
        """Calculate potential field for nodes (intra-group)."""
        intra_potential = {}

        # Calculate mean I value
        i_sum = sum(self.id_to_i.values())
        i_mean = i_sum / len(self.id_to_robots)

        for robot_id in self.id_to_robots.keys():
            robot = self.id_to_robots[robot_id]
            p = PotentialField()

            # Set attractive potential field
            i_value = self.id_to_i[robot_id]
            p.pegra = -self.a * self._gain(i_value - i_mean)

            # Set repulsive potential field
            ro = 0.0
            neighbors = list(self.arc_graph.neighbors(robot_id))

            for neighbor_id in neighbors:
                target_robot = self.id_to_robots[neighbor_id]

                if target_robot.robot_id == robot_id:
                    continue

                if target_robot.group_id != robot.group_id:
                    continue

                if target_robot.fault_a == 1:
                    # Inversely proportional to distance to faulty node
                    weight = self.arc_graph[robot_id][neighbor_id]['weight']
                    ro += 1 / weight

            if robot.fault_a == 1:
                p.perep = float('inf') / 2
            elif ro != 0:
                p.perep = self.b * (self.y * 1 / ro) * (1 / ro)
            else:
                p.perep = 0.0

            intra_potential[robot_id] = p

            # Update overload fault condition
            function = Function(self.id_to_robots, self.id_to_groups)
            fault_o = 1 - function.calculate_over_load_is(self.id_to_robots[robot_id])
            robot.fault_o = fault_o

        return intra_potential

    def calculate_inter_p(self) -> Dict[int, PotentialField]:
        """Calculate potential field for network layer (inter-group)."""
        inter_potential = {}

        for group_id in self.id_to_groups.keys():
            group = self.id_to_groups[group_id]
            p = PotentialField()

            # Calculate attractive potential field for network layer
            p.pegra = self.a * self.xn * group.group_load

            # Calculate repulsive field for network layer
            fk = 0
            robot_id_in_group = group.robot_id_in_group
            for robot_id in robot_id_in_group:
                robot = self.id_to_robots[robot_id]
                if robot.fault_a == 1:
                    fk += 1

            nk = len(robot_id_in_group)
            if fk == nk:
                p.perep = float('inf') / 2
            else:
                p.perep = self.b * (self.yn * fk / (nk - fk))

            inter_potential[group_id] = p

        return inter_potential

    def _gain(self, x: float) -> float:
        """Gain function."""
        return x
